Add implementation of ipfs gateway download

afbjorklund opened the pull request:
## Description of the Pull Request (PR):

Set IPFS_GATEWAY to a shared HTTP gateway, or run `ipfs daemon`.

Upload images like so: `ipfs add --cid-version 1 -w lolcow.sif`.

### This fixes or addresses the following GitHub issues:

 - Fixes #2961

#### Before submitting a PR, make sure you have done the following:

- [x] Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [x] Make sure all commits are signed-off with `git commit -s`, see the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md).
- [x] Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md), if necessary according to the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md).
- [ ] Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)).
- [x] Based this PR against the appropriate branch according to the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md).
- [x] Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md).

DrDaveD commented on internal/pkg/client/net/pull.go:
```diff
@@ -111,8 +111,7 @@ func DownloadImage(ctx context.Context, filePath string, netURL string) error {
 
-// pull will pull a http(s) image into the cache if directTo="", or a specific file if directTo is set.
-func pull(ctx context.Context, imgCache *cache.Handle, directTo, pullFrom string) (imagePath string, err error) {
+func head(pullFrom string) (hash string, err error) {
```
How does it help to introduce this function if it is only used once?

afbjorklund replied:
It was because I wanted to reuse the "pull" code, but in the end I was not able to anyway...

So all it accomplished was to make the functions more similar - but not identical (or shared)

DrDaveD reviewed (changes requested):
Please add the new URL type in appropriate help messages in docs/content.go.

DrDaveD commented:
Also please add a CHANGELOG.md entry.

afbjorklund commented:
Refactored and added a unit test for the the decoding util, from the CIDv1 to the SHA-256...

https://cid.ipfs.tech/#bafybeicn7i3soqdgr7dwnrwytgq4zxy7a5jpkizrvhm5mv6bgjd32wm3q4

When adding the file with --wrap-with-directory, as suggested above, you get two different CID:

https://ipfs.io/ipfs/bafybeicn7i3soqdgr7dwnrwytgq4zxy7a5jpkizrvhm5mv6bgjd32wm3q4/welcome-to-IPFS.jpg

https://ipfs.io/ipfs/bafkreie7ohywtosou76tasm7j63yigtzxe7d5zqus4zu3j6oltvgtibeom?filename=welcome-to-IPFS.jpg

This is similar to how it works in git, when the "tree" gets one SHA-1 and the "blob" gets one SHA-1.

----

In the apptainer implementation, we expect that there is only one SIF file in each directory.

So if you try to download the directory CID, it is created as a symbolic link to the file CID:

```
$ ipfs add --cid-version 1 -w lolcow.sif
added bafybeigsdv6ijdarejvzmx5qwqqpdjsbhuaplpym2nij72dolvmr3i634m lolcow.sif
added bafybeiex3hgq4uslhdpk4iz6ni2oel2o6dt3jqd45lb6vz6ka7krw7rqca 
$ apptainer run ipfs://bafybeiex3hgq4uslhdpk4iz6ni2oel2o6dt3jqd45lb6vz6ka7krw7rqca/lolcow.sif
INFO:    Downloading network image
$ apptainer run ipfs://bafybeigsdv6ijdarejvzmx5qwqqpdjsbhuaplpym2nij72dolvmr3i634m
```

So the second invocation it used the ipfs cache, since it was referring to the _same_ SIF file:

`~/.apptainer/cache/ipfs/97d9cd0e524b38deae233e6a34e22f4ef0e7b4c07ceac3eae7ca07d51b7e3010 -> d21d7c848c11226b965fb0b420f1a6413d00f5bf0cd3509fe86e5d591da3dbe3`
`~/.apptainer/cache/ipfs/d21d7c848c11226b965fb0b420f1a6413d00f5bf0cd3509fe86e5d591da3dbe3`

We can calculate the checksum of the file in the cache, and see that it matches the CID.

Note: https://docs.ipfs.tech/concepts/content-addressing/#cids-are-not-file-hashes

`sha256:1582da68e80fcfc7f7da5cf0679403a89dbd16e10fe7a8258a00b76033d83ffa`
`multi:bafybeigsdv6ijdarejvzmx5qwqqpdjsbhuaplpym2nij72dolvmr3i634m`

But assuming that we **do** use the same encoding and version and codec, it _will_ match:

```
$ ipfs-inspect bafybeigsdv6ijdarejvzmx5qwqqpdjsbhuaplpym2nij72dolvmr3i634m
base32-cidv1-dag-pb-(sha2-256:256:D21D7C848C11226B965FB0B420F1A6413D00F5BF0CD3509FE86E5D591DA3DBE3)
```
